Initial birdhouse prototype.

Zachary-Concordia opened the pull request:
Initial Prototype of birdhouse block, dynamic resource generation, block entity loop, and birdhouse inventory system.
To run: Load the project, create the run configs with gradl, run runData first to generate the .jsons, and then run the client.

Zachary-Concordia committed:
Initial birdhouse prototype.

Initial Prototype of birdhouse block, dynamic resource generation, block entity loop, and birdhouse inventory system.
To run: Load the project, create the run configs with gradl, run runData first to generate the .jsons, and then run the client.

Zachary-Concordia committed:
Simple class diagram.

Class diagram of prototype_1 to show what and where things are.

jamestjw commented on Package Diagram.svg:
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!-- Do not edit this file with editors other than diagrams.net -->
+<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
+<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1" width="1931px" height="1551px" viewBox="-0.5 -0.5 1931 1551" content="&lt;mxfile host=&quot;Electron&quot; modified=&quot;2022-02-17T20:57:28.459Z&quot; agent=&quot;5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) draw.io/16.5.1 Chrome/96.0.4664.110 Electron/16.0.7 Safari/537.3 ...
```
@Zachary-Concordia This file should probably go to the wiki instead

Zachary-Concordia replied:
Added to Wiki

Zachary-Concordia committed:
Deleted and added to Wiki

The content may be found in the wiki here: https://github.com/jamestjw/birdhaus/wiki/Prototype_1-Class-Diagram

jamestjw commented on src/main/java/com/ocelotlovesbirds/birdhaus/datagen/LanguageProviderGen.java:
```diff
@@ -0,0 +1,57 @@
+
+        }
+
+    }
```
what about 

```
 switch (locale) {
   case "fr_ca", "fr_fr" -> addFrenchTranslations();
   default -> addEnglishTranslations();

}
```

implying that we whenever we get some locale that we do not support, we just load the mod using the default language (English)

Zachary-Concordia replied:
Agreed, will change.

jamestjw commented on src/main/java/com/ocelotlovesbirds/birdhaus/datagen/LanguageProviderGen.java:
```diff
@@ -0,0 +1,57 @@
+     */
+    protected void addFrenchTranslations() {
+        add("itemGroup." + ModSetup.TAB_NAME, "Haus de Oiseau");
+        add(Registration.BIRDHOUSE_BLOCK.get(), "Nichoir");
```
I'm curious, what's the difference between these two?

Zachary-Concordia replied:
Was trying to keep the German spelling of house for the tab name to keep the block and mod separate

jamestjw commented on src/main/java/com/ocelotlovesbirds/birdhaus/datagen/LanguageProviderGen.java:
```diff
@@ -0,0 +1,57 @@
+import net.minecraftforge.common.data.LanguageProvider;
+
+public class LanguageProviderGen extends LanguageProvider {
+
+    //TODO: Find a better way to retrieve language.
+    private final String locale;
+
```
```
import java.util.Locale;

Locale.getDefault();
```

would give you something like `en_US`, to see it in action https://code.sololearn.com/cCo0KaHZ9AKj

Zachary-Concordia replied:
Implemented, didn't know about that, thanks!

jamestjw commented on build.gradle:
```diff
@@ -1,22 +1,182 @@
+apply plugin: 'checkstyle'
+
+version = '0.1'
+group = 'team.ocelot.birdhaus' // http://maven.apache.org/guides/mini/guide-naming-conventions.html
```
why not `com.ocelotslovebirds.birdhaus`?

chris7neves replied:
I think version also needs to follow this convention according to forge documentation:
MinecraftVersion-ModMajorVersion.ModMinorVersion.ModPatchVersion

So it would be 1.18-0.1.0

jamestjw replied:
@chris7neves could you drop a link to the documentation

chris7neves replied:
https://mcforge.readthedocs.io/en/latest/conventions/versioning/

Seems like kind of a "nice to have" way to do it.

Here's the paragraph im referring to: "
However, in the case of modding it may be more beneficial to use the format MCVERSION-MAJORMOD.MAJORAPI.MINOR.PATCH to be able to differentiate between world-breaking and API-breaking changes of a mod.
"

jamestjw replied:
im convinced :) i think we should do this

Zachary-Concordia replied:
Also convinced, will change!

jamestjw reviewed (changes requested):
Phew that was a heavy pull request, it took awhile to digest what was going on. I only have a few more comments.

1. Could you fix the structure of the project and the package name 
<img width="392" alt="image" src="https://user-images.githubusercontent.com/44499264/154890490-9a77e8bd-4c9a-42c1-820d-c8e35821d4d1.png">

2. We should really add a space after the slashes when writing comments, something like

```
// A space after the slashes
```

Its super easy to fix this, 
```bash
find src -name '*.java' | xargs perl -pi -e 's/\/\/(\w+)/\/\/ $1/g'
```

Btw you can link this pull request to a task in JIRA by adding the task number to the title of the pull request, something like `[OLB-123123]`

jamestjw commented:
I just tried running the mod, it looks good! Could you mention in the README file how one would run it? It wasn't immediately obvious to me

```
./gradlew runData # Generate data files
./gradlew runClient # Run Minecraft client
```


PS: I think this is better than mentioning click `abc` in `xyz` IDE, it'd obviously not hurt to include how one would do all of this in an IDE as well of course.

jamestjw commented on src/main/java/com/ocelotlovesbirds/birdhaus/datagen/RecipesGen.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    @Override
+    protected void buildCraftingRecipes(Consumer<FinishedRecipe> consumer) {
+        birdhouseRecipe(consumer, Registration.BIRDHOUSE_BLOCK.get(), Blocks.SPRUCE_PLANKS, Blocks.SPRUCE_WOOD);
+    }
```
since the function signature of birdhouseRecipe is `private static void birdhouseRecipe(
             Consumer<FinishedRecipe> recipeConsumer, ItemLike birdhouse, ItemLike log, ItemLike planks)` did you make a mistake with the block types? since the function signature has logs and planks but here you pass in planks and wood @Zachary-Concordia

Zachary-Concordia replied:
Correct this was an error, corrected to logs rather than wood.

shafiqimtiaz reviewed (approved):
Looks good. Tried and it works. Thanks 😄

APLab-ZJ-Projects committed:
Updated based on pull review.

APLab-ZJ-Projects committed:
Updated based on pull review.

APLab-ZJ-Projects committed:
Cleaned up comments to add a space per review request.

Zachary-Concordia reviewed (commented):
Incorporated feedback

shafiqimtiaz committed:
Added Global Tagging (#4)

to enable spawning birdhouse using any type of plank OR logs
